File: 动态规划/cpp/stoneGame.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>

using namespace std;

int maxProfit(const vector<int>& nums)
{
    int n = nums.size();
    vector<vector<int> > dp_table(n, vector<int>(n, 0));

    for(int i=0; i<n; i++)
        dp_table[i][i] = nums[i];

    for(int i=n-2; i>=0; i--)
    {
        for(int j=i+1; j<n; j++)
        {
            dp_table[i][j] = max({nums[i] - dp_table[i+1][j], nums[j]-dp_table[i][j-1]});
        }
    }
    return dp_table[0][n-1];
}

int main()
{
    vector<int> nums = {3, 9, 1, 2};
    cout << maxProfit(nums) << endl;
    return 0;
}

File: DynamicProgramming/python/stoneGame.py
"""
给出几堆石头, 两个人比赛拿石头, 每次只能拿最左边或最右边的的石头堆, 求max(先手拿-后手拿)
"""

import numpy as np


def stone_game(nums):
    if len(nums) == 1:
        return nums[0]
    dp_table = np.zeros((len(nums), len(nums)), dtype=np.int32)
    for i in range(len(nums)):
        dp_table[i, i] = nums[i]

    for i in range(len(nums)-1, -1, -1):
        for j in range(i+1, len(nums)):
            dp_table[i, j] = max(nums[i]-dp_table[i+1, j], nums[j]-dp_table[i, j-1])
    return dp_table[0, -1]


if __name__ == "__main__":
    nums = [3, 9, 1, 2]
    print(stone_game(nums))
